src/proxy/dns.rs

```diff
@@ -0,0 +1,271 @@
+    server_request(&new_message(name, RecordType::A), client_addr, protocol)
+}
+
+/// A short-hand helper for constructing a [Name].
+pub fn n<S: AsRef<str>>(name: S) -> Name {
+    Name::from_utf8(name).unwrap()
+}
+
+/// Helper for parsing a [SocketAddr] string.
+pub fn socket_addr<S: AsRef<str>>(socket_addr: S) -> SocketAddr {
+    socket_addr.as_ref().parse().unwrap()
+}
+
+impl PollingDns {
+    pub(super) async fn new(pi: ProxyInputs, drain: Watch) -> Result<PollingDns, Error> {
+        info!(component = "dns", "dns async polling client started",);
+        Ok(PollingDns {
+            pi,
+            drain,
+            tasks: HashMap::new(),
+        })
+    }
+
+    fn get_handle(
+        proxy_mode: ProxyMode,
+        dns_nameservers: Vec<SocketAddr>,
+        state: Arc<RwLock<ProxyState>>,
+        dns_workload: Workload,
+    ) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            let hostname = dns_workload.hostname.clone();
+            trace!("dns workload async task started for {:?}", &hostname);
+
+            // TODO(kdorosh): don't make a new forwarder for every request?
+            let fw = dns::forwarder_for_mode(proxy_mode, dns_nameservers).unwrap(); // TODO(kdorosh) handle unwrap
+            let r = fw.resolver();
+
+            // Lookup a host.
+            let req = a_request(
+                n(&hostname),
+                socket_addr("1.1.1.1:80"), // TODO(kdorosh): don't hardcode this
+                Protocol::Udp,
+            );
+            let resp = r.lookup(&req).await;
+            if resp.is_err() {
+                warn!(
+                    "dns async response for workload {} is: {:?}",
+                    &dns_workload.uid, resp
+                );
+                return;
+            } else {
+                trace!(
+                    "dns async response for workload {} is: {:?}",
+                    &dns_workload.uid,
+                    resp
+                );
+            }
+            let resp = resp.unwrap();
+            let ips = resp
+                .record_iter()
+                .filter_map(|record| {
+                    if record.rr_type().is_ip_addr() {
+                        // TODO: handle ipv6
+                        return record.data().unwrap().as_a().copied();
+                    }
+                    None
+                })
+                .collect_vec();
+            state
+                .write()
+                .unwrap()
+                .resolved_dns
+                .set_ips_for_workload(dns_workload.uid, ips);
+        })
+    }
+
+    pub(super) async fn run(mut self) {
+        let accept = async move {
+            loop {
+                let dns_workloads = self
+                    .pi
+                    .state
+                    .state
+                    .read()
+                    .unwrap()
+                    .workloads
+                    .get_async_dns_workloads();
```

Comment:
this seems like we are going to resolve for all workloads? Is that right? I thought we were only resolving for workloads that have at least one request recently?

Reply:
yeah that's not right, this is still WIP. im currently reworking the async await part for on-demand and then updating the dns polling to be correct and only poll once per DNS ttl iff we received a request in last DNS ttl lifetime